src/flow/src/repr/relation.rs

```diff
@@ -132,13 +133,24 @@ impl RelationType {
             })
             .try_collect()?;
 
-        let old_to_new_col = BTreeMap::from_iter(
-            mfp.projection
-                .clone()
-                .into_iter()
-                .enumerate()
-                .map(|(new, old)| (old, new)),
-        );
+        let old_to_new_col =
+            BTreeMap::from_iter(mfp.projection.clone().into_iter().enumerate().map(
+                |(new, old)| {
+                    let mut old = old;
+                    // trace back to the original column
+                    while let Some(ScalarExpr::Column(prev)) = if old >= mfp.input_arity {
+                        mfp.expressions.get(old - mfp.input_arity)
+                    } else {
+                        None
+                    } {
+                        old = *prev;
+                        if old < mfp.input_arity {
+                            break;
+                        }
+                    }
+                    (old, new)
+                },
+            ));
```

Comment:
Can you extract it as a function and provide some comments to explain why?

Reply:
> Can you extract it as a function and provide some comments to explain why?

Extracted and added more comments